File: ruby/bob/bob.rb
class Bob
  attr_reader :remark

  def hey(remark)
    @remark = remark
    if yell?
      "Whoa, chill out!"
    elsif question?
      "Sure."
    elsif !has_letters? && !has_numbers?
      "Fine. Be that way!"
    else
      "Whatever."
    end
  end

  private
    def yell?
      has_letters? && remark == remark.upcase
    end

    def question?
      remark[-1] == "?"
    end

    def has_letters?
      remark.chars.any? { |char| char =~ /[[:alpha:]]/ }
    end

    def has_numbers?
      remark.chars.any? { |char| char =~ /[[:digit:]]/ }
    end
end

File: javascript/bob/bob.js
"use strict";

var Bob = function() {};

Bob.prototype.hey = function(input) {
  if(yell(input)) {
    return "Whoa, chill out!";
  } else if(question(input)) {
    return "Sure.";
  } else if(silence(input)) {
    return "Fine. Be that way!";
  } else {
    return "Whatever.";
  }
};

function yell(input) {
  return input == input.toUpperCase() && input.match(/[a-z]/i);
}

function question(input) {
  return input.slice(-1) === "?";
}

function silence(input) {
  return input == 0;
}

module.exports = Bob;
